File: server/app/services/agent_upload/agent_payload_builder.py
"""Agent 输入 payload 构建器.

迁移自 ZHS_Server_java/small/service/agent/AgentPayloadBuilder.java.
"""

from collections import defaultdict
from typing import Any

from app.services.agent_upload.agent_json_helper import normalize, parse_array


def build_input_params(agent: Any, raw_problems: str | None) -> dict[str, Any]:
    """根据 agent 配置的变量和用户 problems 构建输入参数."""
    if agent is None:
        raise ValueError("Agent must not be null")
    safe_problems_json = raw_problems if raw_problems and raw_problems.strip() else "[]"
    problems_array = parse_array(safe_problems_json, "problems")
    variables_in_str = getattr(agent, "agent_variables_in", None) or "[]"
    variables_in_array = parse_array(variables_in_str, "agentVariablesIn")

    param_value_map: dict[str, list[Any]] = defaultdict(list)
    for problem in problems_array:
        target_param = problem.get("parameterName", "")
        if not target_param:
            continue
        variable_obj = next(
            (v for v in variables_in_array if v.get("parameterName") == target_param),
            None,
        )
        raw_content = problem.get("content")
        normalized = normalize(raw_content)
        if normalized is None or normalized == "":
            default = variable_obj.get("default") if variable_obj else None
            if default is not None:
                if isinstance(default, list) and default:
                    normalized = normalize(default[0])
                else:
                    normalized = normalize(default)
        if normalized is None:
            normalized = None
        param_value_map[target_param].append(normalized)

    input_params: dict[str, Any] = {}
    for param_name, values in param_value_map.items():
        if not values:
            input_params[param_name] = None
            continue
        if len(values) == 1:
            input_params[param_name] = values[0]
        else:
            input_params[param_name] = values
    return input_params

File: server/deploy/legacy-archive/java-backend/ZHS_Server_java/src/main/java/com/ai/manager/small/service/agent/AgentPayloadBuilder.java
package com.ai.manager.small.service.agent;

import com.ai.manager.small.domain.AgentUpload;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import org.springframework.stereotype.Component;

import java.util.*;

/**
 * Builds the agent input payload based on the configured variables and user problems.
 */
@Component
public class AgentPayloadBuilder {

    public JsonObject build(AgentUpload agent, String rawProblems) {
        if (agent == null) {
            throw new IllegalArgumentException("Agent must not be null");
        }
        String safeProblemsJson = (rawProblems == null || rawProblems.trim().isEmpty()) ? "[]" : rawProblems;
        JsonArray problemsArray = AgentJsonHelper.parseArray(safeProblemsJson, "problems");
        String variablesInStr = Optional.ofNullable(agent.getAgentVariablesIn()).orElse("[]");
        JsonArray variablesInArray = AgentJsonHelper.parseArray(variablesInStr, "agentVariablesIn");

        JsonObject inputParamsObject = new JsonObject();
        Map<String, List<JsonElement>> paramValueMap = new HashMap<>();

        for (JsonElement element : problemsArray) {
            JsonObject problemObject = element.getAsJsonObject();
            String targetParameterName = problemObject.get("parameterName").getAsString();
            JsonObject variableInObject = findVariable(targetParameterName, variablesInArray);

            JsonElement rawContent = problemObject.get("content");
            JsonElement normalized = AgentJsonHelper.normalize(rawContent);
            if (normalized == null || normalized.isJsonNull()
                    || (normalized.isJsonPrimitive() && normalized.getAsJsonPrimitive().isString()
                    && normalized.getAsString().isEmpty())) {
                JsonElement defEl = variableInObject != null ? variableInObject.get("default") : null;
                if (defEl != null && !defEl.isJsonNull()) {
                    if (defEl.isJsonArray() && defEl.getAsJsonArray().size() > 0) {
                        normalized = AgentJsonHelper.normalize(defEl.getAsJsonArray().get(0));
                    } else {
                        normalized = AgentJsonHelper.normalize(defEl);
                    }
                }
                if (normalized == null) {
                    normalized = JsonNull.INSTANCE;
                }
            }
            paramValueMap.computeIfAbsent(targetParameterName, k -> new ArrayList<>()).add(normalized.deepCopy());
        }

        for (Map.Entry<String, List<JsonElement>> entry : paramValueMap.entrySet()) {
            String paramNameKey = entry.getKey();
            List<JsonElement> values = entry.getValue();
            if (values == null || values.isEmpty()) {
                inputParamsObject.add(paramNameKey, JsonNull.INSTANCE);
                continue;
            }
            if (values.size() == 1) {
                JsonElement single = values.get(0);
                inputParamsObject.add(paramNameKey,
                        (single == null || single.isJsonNull()) ? JsonNull.INSTANCE : single.deepCopy());
                continue;
            }
            JsonArray arr = new JsonArray();
            for (JsonElement v : values) {
                arr.add(v == null || v.isJsonNull() ? JsonNull.INSTANCE : v.deepCopy());
            }
            inputParamsObject.add(paramNameKey, arr);
        }
        return inputParamsObject;
    }

    private JsonObject findVariable(String targetParameterName, JsonArray variablesInArray) {
        for (JsonElement element : variablesInArray) {
            JsonObject obj = element.getAsJsonObject();
            String objParameterName = obj.get("parameterName").getAsString();
            if (targetParameterName.equals(objParameterName)) {
                return obj;
            }
        }
        return null;
    }
}
